### PR TITLE
Ensure static precompiled dependencies are only linked in targets that support linking

### PR DESCRIPTION
**Pull request by kwridan**
Resolves https://github.com/tuist/tuist/issues/4859

### Short description 📝

- Static precompiled targets were currently being "linked" directly by the target that declared a dependency on them
  - note: linking only really took place if the target was one that supports linking (e.g. app, unit tests, etc..), for any static targets the symbols of the precomplied binary symbols would be included
- This worked in simple cases where only one target would depend on the precompiled static target
- In scenarios where multiple static targets needed to depend on the same precompiled static target, build failures would occur due to duplicate symbols
- To solve this, the precompiled static product is now only linked in targets that support linking and instead they are marked as dependencies of the static targets to ensure they can compile
  - For `XCFramework`s those are added to the "Dependencies" copy products build phase
  - For precompiled `Framework`s they are added to the framework search paths _(no change from before the changes)_
  - For precompiled libraries (`.a`) 
     - They are added to the library and swift include search paths of the static targets that depend on them _(no change from before the changes)_
     - Additionally, they are also added to the library search paths of the final linking target

### How to test the changes locally 🧐

XCFramework Test Plan:

- Build this version of tuist

```sh
swift build
```

- Generate the fixture `ios_app_with_xcframeworks`

```sh
swift run tuist generate --path projects/tuist/fixtures/ios_app_with_xcframeworks
```

- Build the project and verify it succeeds
- Verify the precompiled `MyStaticLibrary.xcframework` is only linked in the `App` target
- Verify the precompiled `MyStaticLibrary.xcframework` is declared a dependency (but not in the link phase) of target `StaticFrameworkA`
- Verify `StaticFrameworkA` build successfully in isolation 
- Verify `App` builds and runs successfully without any duplicate symbol errors 

Framework Test Plan:

- Generate the fixture `ios_app_with_static_frameworks`

```sh
swift run tuist generate --path projects/tuist/fixtures/ios_app_with_static_frameworks
```

- Build the project and verify it succeeds
- Verify the precompiled static framework is only linked in the `App` target
- Verify the precompiled static framework is not in the link phase of target `A`
- Verify building the target `A` in isolation still works

### Contributor checklist ✅

- [x] The code has been linted using run `./fourier lint tuist --fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- ~In case the PR introduces changes that affect users, the documentation has been updated~

### Reviewer checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase
- [x] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

**Comment by regularberry** [comment]
@kwridan This is great! It resolves the issue in the fixture I posted here: https://github.com/tuist/tuist/pull/5093 - as well as resolving the issue in our work project. Thank you for this!

**Comment by shahzadmajeed** [comment]
Hi, could we keep following decisions intact?

Currently, static dependencies (both libraries and frameworks) of static frameworks are only linked in app target (or maybe unit tests if they are not testing a host app). This is very useful for keeping apps optimized for size and performance.

**Comment by kwridan** [comment]
> Hi, could we keep following decisions intact?
> 
> Currently, static dependencies (both libraries and frameworks) of static frameworks are only linked in app target (or maybe unit tests if they are not testing a host app). This is very useful for keeping apps optimized for size and performance.

The changes in this PR don't change that rule :) - if anything, it further enforces it for pre-complied frameworks and libraries.

File: Sources/TuistCore/Graph/GraphTraverser.swift [hunk 1]
```diff
@@ -232,6 +232,9 @@ public class GraphTraverser: GraphTraversing {
 
     public func searchablePathDependencies(path: AbsolutePath, name: String) throws -> Set<GraphDependencyReference> {
         try linkableDependencies(path: path, name: name, shouldExcludeHostAppDependencies: false)
+            .union(
+                staticPrecompiledDependenciesAndTheirDependencies(path: path, name: name)
+            )
```
Comment (danyf90) [inline Sources/TuistCore/Graph/GraphTraverser.swift:237]: nit
```suggestion
            .union(staticPrecompiledDependenciesAndTheirDependencies(path: path, name: name))
```

**Review by danyf90 — APPROVED** [review]

File: Sources/TuistCore/Graph/GraphTraverser.swift (hunk 1 above)
Comment (kwridan) [reply Sources/TuistCore/Graph/GraphTraverser.swift]: applied in https://github.com/tuist/tuist/pull/5107/commits/61f6762d26191e6b0db2c7f8c4bf350784b47b0a

**Comment by kwridan** [comment]
pushed the fixture update to include the issue in https://github.com/tuist/tuist/issues/4859

**Comment by kwridan** [comment]
@danyf90 can we add @regularberry as a contributor so that he can submit a review to unlock to final check? We need another reviewer for this PR and aside than both of us and @fortmarek not sure if anyone else has sufficient context to review this PR

thanks

**Review by regularberry — APPROVED** [review]
Oh, I can review! This works great, and is a more holistic solution than the PR I posted. Nice job 👍🏻

**Review by danyf90 — APPROVED** [review]

**Comment by kwridan** [comment]
thanks all!

**Comment by regularberry** [comment]
@kwridan  Thanks for getting this in! Do you know how the process works for cutting a new version? Is it on a regular cadence? We'd love to pull this fix into our project

**Pull request merged**